array.c

```diff
@@ -3554,55 +3674,99 @@ static VALUE rb_ary_bsearch_index(VALUE ary);
- *  that satisfies the block), this method returns nil.
- *
- *     ary = [0, 4, 7, 10, 12]
- *     # try to find v such that 4 <= v < 8
- *     ary.bsearch {|x| 1 - x / 4 } #=> 4 or 7
- *     # try to find v such that 8 <= v < 10
- *     ary.bsearch {|x| 4 - x / 2 } #=> nil
- *
- *  You must not mix the two modes at a time; the block must always
- *  return either true/false, or always return a number.  It is
- *  undefined which value is actually picked up at each iteration.
+ *    array.bsearch {|element| ... } -> object
+ *    array.bsearch -> new_enumerator
+ *
+ *  Returns an element from +self+ selected by a binary search.
+ *  +self+ should be sorted, but this is not checked.
+ *
+ *  There are two search modes:
+ *  - <b>Find-minimum mode</b>: the block should return +true+ or +false+.
+ *  - <b>Find-any mode</b>: the block should return a numeric value.
+ *
+ *  The block should not mix the modes by and sometimes returning +true+ or +false+
+ *  and sometimes returning a numeric value, but this is not checked.
+ *
+ *  ====== Find-Minimum Mode
+ *
+ *  In find-minimum mode, the block always returns +true+ or +false+.
+ *  The further requirement (though not checked) is that
+ *  there are no indexes +i+ and +j+ such that:
+ *  - <tt>0 <= i < j <= self.size</tt>.
+ *  - The block returns +true+ for <tt>self[i]</tt> and +false+ for <tt>self[j]</tt>.
+ *
+ *  In find-minimum mode, method bsearch returns the first element for which the block returns true.
+ *
+ *  Examples:
+ *    a = [0, 4, 7, 10, 12]
+ *    a.bsearch {|x| x >= 4 } # => 4
+ *    a.bsearch {|x| x >= 6 } # => 7
+ *    a.bsearch {|x| x >= -1 } # => 0
+ *    a.bsearch {|x| x >= 100 } # => nil
+ *
+ *  Less formally: the block is such that all +false+-evaluating elements
+ *  precede all +true+-evaluating elements.
+ *
+ *  These make sense as blocks in find-minimum mode:
+ *    a = [0, 4, 7, 10, 12]
+ *    a.map {|x| x >= 4 } # => [false, true, true, true, true]
+ *    a.map {|x| x >= 6 } # => [false, false, true, true, true]
+ *    a.map {|x| x >= -1 } # => [true, true, true, true, true]
+ *    a.map {|x| x >= 100 } # => [false, false, false, false, false]
+ *
+ *  This would not make sense:
+ *    a = [0, 4, 7, 10, 12]
+ *    a.map {|x| x == 7 } # => [false, false, true, false, false]
+ *
+ *  ====== Find-Any Mode
+ *
+ *  In find-any mode, the block always returns a numeric value.
+ *  The further requirement (though not checked) is that
+ *  there are no indexes +i+ and +j+ such that:
+ *  - <tt>0 <= i < j <= self.size</tt>.
+ *  - The block returns a negative value for <tt>self[i]</tt>
+ *    and a positive value for <tt>self[j]</tt>.
+ *  - The block returns a negative value for <tt>self[i]</tt> and zero <tt>self[j]</tt>.
+ *  - The block returns zero for <tt>self[i]</tt> and a positive value for <tt>self[j]</tt>.
+ *
+ *  In find-any mode, method bsearch returns some element
+ *  for which the block returns zero, or +nil+> if no such element is found.
```

Comment:
It seems the `>` is a typo that snuck in, maybe it should be `,`?

Reply:
Fixed.